### PR TITLE
plutus-tx: compile let-bindings of non-values lazily

### PR DESCRIPTION
**Pull request by michaelpj**
Consider `let e = error in if x then e else ()`. This is conditionally error in Haskell, but unconditionally error in PLC.

The user can write such things, or (more problematically) GHC can generate them by lifting things out. This happened when I was working on the `plutus-use-cases` code, so it's a real worry.

We can solve this by compiling these let-bindings lazily. But which ones should we do this for? The best heuristic I've been able to come up with is the same as for case branches: if it's not a value. This is overly aggressive, since it will ding "fine" things like `let x = 1+1 in x + 1`, but I don't have a better check right now.

The evaluation test I included fails without this patch, and it also fixes the problems I had with `plutus-use-cases`.

I had to add some state to the Plutus Tx compiler, so I hid the `MonadState` instances for `QuoteT` and `DefT`, which expose internal state anyway.

### CODE REVIEW

**Comment by j-mueller** [comment]
Which of the contacts in `plutus-use-cases` caused those problems?

**Comment by michaelpj** [comment]
It's the state machine stuff. I don't really know why, but GHC decided to pull out the error calls there where it hadn't before. It's definitely not the contract's fault, this is easy to trigger otherwise, we just somehow hadn't hit it before.

**Comment by michaelpj** [comment]
Indeed, this got a lot easier to work on once I managed to get a test case that didn't involve the whole of `plutus-use-cases` :sweat_smile:

**Comment by michaelpj** [comment]
Also cc @effectfully who was annoyed by the `MonadState` instance for `QuoteT` - this removes it.

**Comment by michaelpj** [comment]
@j-mueller here's the final commit removing stuff from `plutus-use-cases` - it was only the very final bit where this showed up! https://github.com/michaelpj/plutus/commit/bb18b9f19a67abfc7a78773126c72242bc7a239b

**Comment by j-mueller** [comment]
> It's the state machine stuff

Thanks, I asked because I am removing TH from the state machine bits at the moment and saw that some of the tests started to fail. So presumably it's that.

**Comment by michaelpj** [comment]
> Thanks, I asked because I am removing TH from the state machine bits at the moment and saw that some of the tests started to fail. So presumably it's that.

Exactly. I've been banging my head against that since my last PR ;)

File: language-plutus-core/src/Language/PlutusCore/Quote.hs [hunk 1]
```diff
@@ -51,7 +53,13 @@ emptyFreshState = Unique 0
 -- | The "quotation" monad transformer. Within this monad you can do safe construction of PLC terms using quasiquotation,
 -- fresh-name generation, and parsing.
 newtype QuoteT m a = QuoteT { unQuoteT :: StateT FreshState m a }
-    deriving (Functor, Applicative, Monad, MonadTrans, MM.MFunctor, MonadState FreshState, MonadError e, MonadReader r)
+    deriving (Functor, Applicative, Monad, MonadTrans, MM.MFunctor, MonadError e, MonadReader r)
+
+-- Need to write this by hand, deriving wants to derive the one for DefState
+instance MonadState s m => MonadState s (QuoteT m) where
```
Comment (effectfully) [inline language-plutus-core/src/Language/PlutusCore/Quote.hs]: Great. I wonder whether we can use `DerivingVia` here somehow.

**Pull request merged**